### PR TITLE
Feat/119 slicing advantage score

### LINKED ISSUE #119: Implement slicing as in Tensorflow Membership privacy attack
* Slicing correctly classified vs wrongly classified
* Slicing per class
* Output: should give attack advantage score for the whole dataset, and for each slice individually
* Have slices as optional arguments (if nothing is specified when running the membership inference attack, then run it on whole dataset). It should also be possible to specify several in one time, which would run attack on every slice

* sidenote: run tensorflow's attack and learn from it!

### PR DESCRIPTION
**Pull request by jtorhoff**
Resolves #119. 

@blauertee sorry I renamed the branch which automatically closed the pull request you had created :\

### CODE REVIEW

**Comment by jtorhoff** [comment]
@marisanest I pushed the notebook which uses the new analysis class into the repo but you can also access it here: https://colab.research.google.com/drive/11kCsn36bjHYzVutthEjyHsvho5zJBqOX?usp=sharing

it uses the pytorch model.

The other notebook which uses tf privacy works with the tf model and you can access it here:
https://colab.research.google.com/drive/1hC1lDMBjthhJhXiuqXcnvYqHfk6qpdkG?usp=sharing

I wanted to share the kaggle notebooks but they require e-mails so I copied them over to Colab.

File: privacy_evaluator/attacks/membership_inference/membership_inference_analysis.py
Comment (marisanest) [inline privacy_evaluator/attacks/membership_inference/membership_inference_analysis.py:75]: Maybe it would make sense to not `__init__` and `fit` a new attack for every analysis. You could also provide the `target_model` as a  `__init__` parameter for the `MembershipInferenceAttackAnalysis` class, so that the analysis is always related to the same `target_model` and `attack_type`.  You could then check via the `_art_attack_model_fitted` parameter of the attack if the attack is already fitted and only if not you fit it.

Sorry for the long text, hopefully, it is understandable. It is also just a more subjective suggestion, not needed for approval.

File: privacy_evaluator/attacks/membership_inference/membership_inference_analysis.py
Comment (marisanest) [inline privacy_evaluator/attacks/membership_inference/membership_inference_analysis.py:14]: Maybe it would make sense to put this `@datacass` under the `data_structure` folder.

File: privacy_evaluator/attacks/membership_inference/membership_inference_analysis.py
Comment (marisanest) [inline privacy_evaluator/attacks/membership_inference/membership_inference_analysis.py:87]: I would suggest putting both metrics into our metrics module (even though you just use the `roc_curve` from `sklearn`) and you maybe don't even need the `roc_curve` if you only need the `tpr` and `fpr` values.

**Review by marisanest — APPROVED** [review]
Looks great and works just perfect :tada: 

I just added some small suggestions but not needed for approval. Also, one or two tests would be greate to have.

File: privacy_evaluator/attacks/membership_inference/membership_inference_analysis.py
Comment (jtorhoff) [reply privacy_evaluator/attacks/membership_inference/membership_inference_analysis.py:75]: Yes that makes sense and is an interesting suggestion. I'll do that later today.

File: privacy_evaluator/attacks/membership_inference/membership_inference_analysis.py
Comment (jtorhoff) [reply privacy_evaluator/attacks/membership_inference/membership_inference_analysis.py:14]: I think it's a matter of preference. This particular data class is only used and will only ever be used by the `MembershipInferenceAttackAnalysis` class so for me it makes more sense to keep them together. Maybe we can wait until the sprint review and then see what our stakeholders have to say.

File: privacy_evaluator/attacks/membership_inference/membership_inference_analysis.py
Comment (jtorhoff) [reply privacy_evaluator/attacks/membership_inference/membership_inference_analysis.py:87]: Makes sense, I will do that.

**Comment by jtorhoff** [comment]
I will try to add some tests, thanks for the detailed review :)

**Comment by marisanest** [comment]
Ok sure. I will merge this PR for now, because I need the changes in order to update the notebooks. If you or someone else want to change or add something, we can open a new PR :)

**Pull request merged**
